breathecode/certificate/views.py

```diff
@@ -25,10 +26,38 @@
 
 @api_view(["GET"])
 @permission_classes([AllowAny])
-def get_specialties(request):
-    items = Specialty.objects.all()
-    serializer = SpecialtySerializer(items, many=True)
-    return Response(serializer.data, status=status.HTTP_200_OK)
+@capable_of("read_certificate")
+def get_academy_specialties(request, academy_id=None):
+    academy_id = request.headers.get("Academy") or request.headers.get("academy")
+    if not academy_id or not academy_id.isdigit():
+        return Response({"detail": "Valid academy ID is required"}, status=status.HTTP_400_BAD_REQUEST)
+    academy_id = int(academy_id)
+
+    items = Specialty.objects.filter(syllabus__academy_owner=academy_id).distinct()
+
+    like = request.GET.get("like")
+    if like:
+        items = items.filter(Q(name__icontains=like) | Q(syllabus__name__icontains=like))
+
+    syllabus_slug = request.GET.get("syllabus_slug")
+    if syllabus_slug:
+        items = items.filter(syllabus__slug=syllabus_slug).distinct()
+
+    allowed_sort_fields = ["created_at", "-created_at", "name", "-name"]
+    sort = request.GET.get("sort", "-created_at")
+    if sort not in allowed_sort_fields:
+        return Response({"detail": "Invalid sort field"}, status=status.HTTP_400_BAD_REQUEST)
+    items = items.order_by(sort)
+
+    paginator = HeaderLimitOffsetPagination()
+    page = paginator.paginate_queryset(items, request)
+    serializer = SpecialtySerializer(page, many=True)
+
+    return (
+        paginator.get_paginated_response(serializer.data)
+        if paginator.is_paginate(request)
+        else Response(serializer.data, status=status.HTTP_200_OK)
+    )
```

Comment:
It should look more like this:

```py
if self.is_paginate(request):
    return self.get_paginated_response(serializer.data)
        
return Response(serializer.data, status=status.HTTP_200_OK)
```

Reply:
In the modified code, self is not used because we are not within a class-based APIView, but rather a function-based view with @api_view. In a class, self is used to access instance methods and attributes, as in the case of CertificateAcademyView. However, in a function-based view, self is not necessary since we are not working with a class instance. Pagination methods are called directly from the HeaderLimitOffsetPagination instance.

Reply:
In the modified code, self is not used because we are not within a class-based APIView, but rather a function-based view with [@api](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html)_view. In a class, self is used to access instance methods and attributes, as in the case of CertificateAcademyView. However, in a function-based view, self is not necessary since we are not working with a class instance. Pagination methods are called directly from the HeaderLimitOffsetPagination instance.